internal/pipeline/types_test.go

```diff
@@ -22,170 +22,190 @@ import (
 	"sigs.k8s.io/kustomize/kyaml/yaml"
 )
 
-func TestPipelineParse(t *testing.T) {
-	type testcase struct {
-		Input    string
-		Expected Pipeline
-		Error    bool
+func TestString(t *testing.T) {
+	expected := `apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+    name: pipeline
+sources:
+  - ./*
+`
+	actual, err := NewPipeline().String()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	if !assert.EqualValues(t, expected, actual) {
+		t.Fatalf("unexpected string value")
 	}
+}
 
-	testcases := []testcase{
+var expected Pipeline = Pipeline{
+	Sources: []string{
+		"a",
+		"b",
+	},
+	Generators: []Function{
+		{
+			Image: "gcr.io/kpt-functions/generate-folders",
+			Config: *yaml.MustParse(`apiVersion: cft.dev/v1alpha1
+kind: ResourceHierarchy
+metadata:
+name: root-hierarchy
+namespace: hierarchy # {"$kpt-set":"namespace"}`).YNode(),
+		},
+	},
+	Transformers: []Function{
+		{
+			Image:      "patch-strategic-merge",
+			ConfigPath: "./patch.yaml",
+		},
 		{
-			Input: `
+			Image: "gcr.io/kpt-functions/set-annotation",
+			ConfigMap: map[string]string{
+				"environment": "dev",
+			},
+		},
+	},
+	Validators: []Function{
+		{
+			Image: "gcr.io/kpt-functions/policy-controller-validate",
+		},
+	},
+}
+
+func TestAdd(t *testing.T) {
+	actual := (&Pipeline{}).
+		AddSources("a", "b").
+		AddGenerators(expected.Generators...).
+		AddTransformers(expected.Transformers...).
+		AddValidators(expected.Validators...)
+
+	if !isPipelineEqual(t, *actual, expected) {
+		t.Fatalf("build result is different from expected")
+	}
+}
+
+func TestSet(t *testing.T) {
+	actual := NewPipeline().
+		SetName("").
+		SetKind("").
+		SetAPIVersion("").
+		SetSources([]string{"a", "b"}).
+		SetGenerators(expected.Generators).
+		SetTransformers(expected.Transformers).
+		SetValidators(expected.Validators)
+	if !isPipelineEqual(t, *actual, expected) {
+		t.Fatalf("build result is different from expected")
+	}
+}
+
+type testcase struct {
+	Input    string
+	Expected Pipeline
+	Error    bool
+}
+
+var testcases map[string]testcase = map[string]testcase{
+	"simple": {
+		Input: `
 apiVersion: kpt.dev/v1alpha1
 kind: Pipeline
 metadata:
   name: pipeline
 `,
-			Expected: Pipeline{
-				ResourceMeta: yaml.ResourceMeta{
-					TypeMeta: yaml.TypeMeta{
-						APIVersion: "kpt.dev/v1alpha1",
-						Kind:       "Pipeline",
-					},
-					ObjectMeta: yaml.ObjectMeta{
-						NameMeta: yaml.NameMeta{
-							Name: "pipeline",
-						},
-					},
-				},
-			},
-		},
-		{
-			Input: `
+		Expected: *NewPipeline().
+			SetAPIVersion("kpt.dev/v1alpha1").
+			SetKind("Pipeline").
+			SetName("pipeline").
+			SetSources(nil),
```

Comment:
why is this nil instead of default?

Reply:
Because if there is no `sources` in the yaml input, the `Sources` slice will be `nil`. `nil` is different from empty slice.

Reply:
I'd expected this would default to './*'.